src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+from typing import cast
+
+from pants.backend.python.subsystems.python_tool_base import PythonToolBase
+from pants.backend.python.target_types import ConsoleScript, PythonRequirementsField
+from pants.backend.python.util_rules.interpreter_constraints import InterpreterConstraints
+from pants.backend.python.util_rules.pex import PexRequest, PexRequirements, VenvPex, VenvPexProcess
+from pants.engine.fs import CreateDigest, Digest, FileContent, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.process import ProcessResult
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule
+from pants.engine.target import Targets
+from pants.util.strutil import pluralize
+
+logger = logging.getLogger(__name__)
+
+
+class PipToolsSubsystem(PythonToolBase):
+    options_scope = "pip-tools"
+    help = "Used to generate lockfiles for third-party Python dependencies."
+
+    default_version = "pip-tools==6.2.0"
+    default_main = ConsoleScript("pip-compile")
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        # TODO: How should users indicate where to save the lockfile to when we have per-tool
+        #  lockfiles and mmultiple user lockfiles?
```

Comment:
One approach would be to have a `resolve(name=..)` target type, with optional parameters for overriding some default lockfile templates. To use it, you would declare one in some directory, and then when it was actually consumed by a `*_binary` target, the lockfiles would be produced next to it at the templated paths (templated by platform * interpreter_constraint * etc).

A downside of that modeling though is that it means that whatever our "default" global resolve is would be a bit magical. We don't want folks to have to declare one of these just to get started with Pants. Not sure how to fix that mismatch... maybe with a setting instead...?: ```[resolves] resolve_mapping = "{'default': '3rdparty/python/lock%p-%i', 'custom_for_aws': '3rdparty/python/aws_lock%p-%i'}"```.